File: 7/day7.py
def main(file_path: str) -> None:
    def can_obtain(target: int, arr: list[int]) -> bool:
        print(f'{target=}, {arr=}')
        if len(arr) == 1:
            return target == arr[0]
        # target divisible by the last number
        if target % arr[-1] == 0 and can_obtain(target // arr[-1], arr[:-1]):
            return True
        if target > arr[-1] and can_obtain(target - arr[-1], arr[:-1]):
            return True
        # For part 2
        # s_target = str(target)
        # s_last = str(arr[-1])
        # if len(s_target) > len(s_last) and s_target.endswith(s_last) and \
        #     can_obtain(int(s_target[:-len(s_last)]), arr[:-1]):
        #         return True
        return False
    
    total = 0
    with open(file_path) as f:
        for line in f:
            part1, part2 = line.split(': ')
            target = int(part1)
            arr = list(map(int, part2.split()))
            if can_obtain(target, arr):
                total += target
    print(total)
       
if __name__ == "__main__":
    main("./7/input0.txt")
    # main("./7/input.txt")

File: 7/day7.js
const fs = require('fs');

function canObtain(target, arr) {

    if (arr.length === 1) {
        return target === arr[0];
    }
    const last = arr[arr.length - 1];
    let subarr = arr.slice(0, -1);
    if (target % last === 0 && canObtain(Math.floor(target / last), subarr)) {
        return true;
    }

    if (target > last && canObtain(target - last, subarr)) {
        return true;
    }

    const sTarget = target.toString();
    const sLast = last.toString();
    if (sTarget.length > sLast.length && sTarget.endsWith(sLast) && canObtain(parseInt(sTarget.slice(0, -sLast.length)), arr.slice(0, -1))) {
        return true;
    }
    return false;
}

function main(filePath) {
    let total = 0;
    const data = fs.readFileSync(filePath, 'utf8').split('\n');
    data.forEach(line => {
        if (line === '') return;
        const [part1, part2] = line.split(': ');
        const target = parseInt(part1);
        const arr = part2.split(' ').map(Number);
        if (canObtain(target, arr)) {
            total += target;
        }
    });

    console.log(total);
}

main("./7/input0.txt");
main("./7/input.txt");
